docs/proposals/scalability-testing.md

```diff
@@ -0,0 +1,78 @@
+
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/scalability-testing.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+## Background
+
+We have a goal to be able to scale to 1000-node clusters by end of 2015.
+However, running regression tests on 1000-node clusters or requiring developers
+to test their changes (with respect to performance) on 1000-node clusters is
+ridiculous. So we need some infrastructure to simulate big clusters on relatively
+small number of machines.
+
+
+## Proposal
+
+In this proposal we are focusing on scalability testing of the whole cluster.
+We do NOT focus on node-scalability - this issue should be handled separately.
+
+Since we do not focus on the node performance, we don't need real Kubelet and
+KubeProxy - in fact we don't even need to start real containers.
+All we actually need is to have some Kubelet-like and KubeProxy-like components
+that will be simulating the load on apiserver that their real equivalents are
+generating (e.g. sending NodeStatus updated, watching for pods, watching for
+endpoints (KubeProxy), etc.).
+
+What needs to be done:
+1. Determine what requests both KubeProxy and Kubelet are sending to apiserver.
+2. Create a KubeletSim that is generating the same load on apiserver that the
+   real Kubelet, but is not starting any containers. In the initial version we
+   can assume that pods never die, so it is enough to just react on the state
+   changes read from apiserver.
+	 TBD: Maybe we can reuse a real Kubelet for it by just injecting some "fake"
+   interfaces to it?
```

Comment:
Let's not get ahead of ourselves. I don't think we should think how it will be packed into binaries, or set up, but how to write a library that can simulate a single node.